### PR TITLE
Refactor IC3 implementations to common, parameterized structure

### PR DESCRIPTION
**Pull request by makaimann**

### CODE REVIEW

**Comment by makaimann** [comment]
Hey @ahmed-irfan, I thought I'd make a draft PR where we can discuss the IC3 updates. What do you think of this base class? I started with templates but it was a bit of a pain and I thought maybe we didn't really need it.

Also, I remembered why we had frames over terms instead of the main unit. It's because Frame 0 (e.g. initial states) might not be in that form. We could shift all the frames and never store init in the structure. Or we could just use frames over terms. What do you think?

**Comment by ahmed-irfan** [comment]
> Hey @ahmed-irfan, I thought I'd make a draft PR where we can discuss the IC3 updates. What do you think of this base class? I started with templates but it was a bit of a pain and I thought maybe we didn't really need it.
> 
> Also, I remembered why we had frames over terms instead of the main unit. It's because Frame 0 (e.g. initial states) might not be in that form. We could shift all the frames and never store init in the structure. Or we could just use frames over terms. What do you think?

I think it looks good

File: engines/ic3base.h
Comment (zhanghongce) [inline engines/ic3base.h:83]: Why not `const std::shared_ptr<IC3Goal> &`? So you can save one copy in the argument.

Comment (makaimann) [reply engines/ic3base.h]: Great catch! That must have been overlooked because it was a unique pointer at one point that got moved instead of copied. Addressed your comment with d3bd2b4959000b9782a5e431d007b4ae0d549a39

**Comment by makaimann** [comment]
All right, I finished moving all the IC3 implementations to this new infrastructure so I think we're ready for reviews. There are still TODOs in the code, some of which we can address now and others can probably be a separate PR later. I think it would be good to evaluate this against the `hwmcc2020` branch and maybe `master` as well since it's a big change, so I've marked this `do not merge` for now. I'll start evaluation runs tomorrow.

File: engines/ic3.cpp
Comment (makaimann) [inline engines/ic3.cpp:213]: @ahmed-irfan flagging this TODO for you. Do you think it would be a cleaner interface for `inductive_generalization` to take a disjunction instead of a conjunction? E.g. take something that's already aligned with the frame and generalize it.

File: engines/ic3.cpp
Comment (makaimann) [inline engines/ic3.cpp:223]: My point here is it would be nice if we didn't have these assumptions about the solver state. I didn't want someone implementing the derived class to have to pop the solver context, so that's done after calling this method instead of in it. But that also means that you can't use the solver in `generalize_predecessor` because it's still at context level 1 and has assertions from the relative inductiveness check.

File: engines/ic3base.h
Comment (makaimann) [inline engines/ic3base.h:75]: To avoid over-generalizing, decided to just assume that an `IC3Formula` is always a disjunction or conjunction. The advantage of this is that it allows more work to be done in the base class instead of relying on the derived classes to do it. For example, this allows for using the unsat core for a first round of generalization if you know the structure. This could be changed later.

File: engines/prover.h
Comment (makaimann) [inline engines/prover.h:32]: These are just supposed to be used by CEGAR loops to communicate how refinement went. In IC3Base, `refine` will return `REFINE_NONE` by default -- meaning no refinement is necessary, so a counterexample is always real unless this is overridden.

File: options/options.cpp
Comment (makaimann) [inline options/options.cpp:52]: Changed this because that option is only used in `mbic3`.

**Review by makaimann — COMMENTED** [review]
Left a few comments to point out and explain some things.

File: engines/ic3.cpp
Comment (ahmed-irfan) [inline engines/ic3.cpp:119]:
```suggestion
//TODO: use unsatcore-reducer
```

File: engines/ic3.cpp
Comment (ahmed-irfan) [reply engines/ic3.cpp:193]: i see most algorithms written in papers use cube as input. So, i would stick to that notation. 
Also, imagine, if we don't do any generalization, then this function would return the negation of the cube.

File: engines/ic3ia.h
Comment (ahmed-irfan) [inline engines/ic3ia.h:46]: add empty line space

File: engines/ic3ia.h
Comment (ahmed-irfan) [inline engines/ic3ia.h:49]: add  empty line space

File: engines/ic3ia.h
Comment (ahmed-irfan) [inline engines/ic3ia.h:53]: empty line space

File: engines/ic3ia.h
Comment (ahmed-irfan) [inline engines/ic3ia.h:57]: empty line space

File: engines/ic3ia.cpp
Comment (ahmed-irfan) [inline engines/ic3ia.cpp:213]: it shouldn't be a problem, I think.  or Maybe I am missing something here.

File: engines/ceg_prophecy_arrays.cpp
Comment (ahmed-irfan) [inline engines/ceg_prophecy_arrays.cpp:118]:
```suggestion
//TODO : think about making it use the same prover -- incrementally
    shared_ptr<Prover> prover =
```

**Review by ahmed-irfan — CHANGES_REQUESTED** [review]
Ok. Finally, I managed to do the review. Thanks for being patient :)

The code looks really cool now. 

I left some comments -- don't need to make all the changes, but we should keep the notes / todos

File: engines/ic3base.h
Comment (makaimann) [reply engines/ic3base.h:187]: What do you think of just `get_model_ic3_formula`? I only say that because they're not necessarily equalities.

File: engines/ic3base.h
Comment (makaimann) [reply engines/ic3base.h:152]: Good point! I'll look into that.

File: engines/ic3base.h
Comment (makaimann) [reply engines/ic3base.h:74]: Ahh yeah that would be nice, but then we'd need a solver in `IC3Formula`. Do you think it's worth it?

Comment (ahmed-irfan) [reply engines/ic3base.h:74]: you have a good point. Let's not do it then.

File: engines/ic3ia.h
Comment (makaimann) [reply engines/ic3ia.h:51]: done in 776b789b56bedbe54bb965348ada5bd2621dd658

File: engines/ic3base.h
Comment (ahmed-irfan) [reply engines/ic3base.h]: sounds good

File: engines/ic3ia.h
Comment (makaimann) [reply engines/ic3ia.h:56]: done in 776b789b56bedbe54bb965348ada5bd2621dd658

File: engines/ic3ia.h
Comment (makaimann) [reply engines/ic3ia.h:61]: done in 776b789b56bedbe54bb965348ada5bd2621dd658

File: engines/ic3ia.cpp
Comment (makaimann) [reply engines/ic3ia.cpp:215]: So it was failing, e.g. the formula wasn't unsatisfiable. I think it has something to do with the predicate abstraction? I need to look into it more still. Might be why the MathSAT IC3IA has a bail-out option from the generalization. Even the version for non-deterministic systems isn't guaranteed to be unsat I don't think because we're not using model values.

File: engines/ic3base.h
Comment (makaimann) [reply engines/ic3base.h]: Done with a96e40470732592a903a69f6b4a8050abd28ad9d

File: engines/ic3base.cpp
Comment (makaimann) [reply engines/ic3base.cpp]: Done with 8d6d931f26c785b86ac11b81bebdf59efe3ebcd3

File: engines/ic3base.h
Comment (makaimann) [reply engines/ic3base.h]: Fixed with https://github.com/upscale-project/pono/pull/122/commits/efb4217f60e9fd949c18b32c8db88a2ae636e3e8

File: engines/ic3base.h
Comment (makaimann) [reply engines/ic3base.h]: Fixed with https://github.com/upscale-project/pono/pull/122/commits/efb4217f60e9fd949c18b32c8db88a2ae636e3e8

File: engines/ic3base.cpp
Comment (makaimann) [reply engines/ic3base.cpp]: Done with https://github.com/upscale-project/pono/pull/122/commits/141f0bc5704f038b5b1f495ac7fc6e6fdc6d24aa

File: engines/ic3base.cpp
Comment (makaimann) [reply engines/ic3base.cpp]: Good catch -- removed with https://github.com/upscale-project/pono/pull/122/commits/b70823491c951eb76c679501f3bdac9ff33e6470

File: engines/ic3base.cpp
Comment (makaimann) [reply engines/ic3base.cpp]: Done with https://github.com/upscale-project/pono/pull/122/commits/2f153be595b04c432876af62445ae23fd0f60bc9

File: engines/ic3base.cpp
Comment (makaimann) [reply engines/ic3base.cpp]: Good idea! Done with https://github.com/upscale-project/pono/pull/122/commits/bf4ba8a1d350a51f2da4e85a1a036c3b5b971fdf

File: engines/ic3base.h
Comment (makaimann) [reply engines/ic3base.h:153]: So if we use the `reducer_` for `inductive_generalization` in `IC3` and `ModelBasedIC3`, would we need to get rid of this https://github.com/makaimann/smt-switch/blob/6e6552140d667e02d3ed2ae7b08b3902af051f2b/src/utils.cpp#L206? Cause for inductive generalization, I think you want to drop a literal, then reduce with an unsat core and repeat. But when you drop a literal, it's not guaranteed to be unsat, right?

**Comment by makaimann** [comment]
@ahmed-irfan here's the initial comparison. There are no discrepancies at least but there are still some fairly major issues:

1. There are a bunch of errors in mbic3 which I will look into
2. Our internal ic3ia is much worse than the external one. Maybe we can look into this more together? Do you have any intuition on why that might be? I can think of three possibilities.
  * there's an issue in our bit-level IC3 algorithm
  * we didn't set the right options in mathsat (because we didn't set any besides unsat-cores and models)
  * overhead from smt-switch

param-ic3-comp

**Comment by makaimann** [comment]
I looked into the errors in `mbic3` and that might be related to the poor performance in `ic3ia` as well. That bug should be fixed by dc27b623a08816d1a98e9cf4c707ed2f110fac33. I'll start new jobs soon.

**Comment by makaimann** [comment]
We want to merge this once we've confirmed that the bug is fixed (cluster currently down). And then we can make improvements in a separate PR. We need to focus on improving the ic3-flavor algorithms' performance. We discussed offline some possible causes for slowdowns:
* bug that was just fixed in `propagate` / `constrain_frame`
* using unsat core reducer with default 0 --> e.g. minimal core. in generalize_predecessor could use 2
* options for the `msat_env`
  * could try getting `msat_env` from `msat-ic3ia` so that the configuration is the same
  * could even use a shared env to avoid translation
* doing interpolation outside instead of using sequence interpolant (our outside loop could be more expensive because it needs to re-solve each time)
  * and do we even need sequence interpolants at all

File: engines/ic3ia.cpp
Comment (ahmed-irfan) [inline engines/ic3ia.cpp:332]: this is going to add the predicates before doing the reduction. We don't want that.

Comment (ahmed-irfan) [reply engines/ic3ia.cpp]: fixed in 31b219f0ea644c93f268b02052895575b6fe765c

Comment (makaimann) [reply engines/ic3ia.cpp]: Oops, thanks for catching that.

**Comment by makaimann** [comment]
Found another issue, just fyi @ahmed-irfan. So `ic3ia` doesn't work correctly with `--static-coi`. This is because `ConeOfInfluence` reduction is done in the `Prover::initialize`. However, we compute the abstraction before calling initialize so that we can pass the abstract system as the main `TransitionSystem`. In theory this would just work, but because all the constraints for implicit predicate abstraction, none of the variables can be removed. This should definitely be fixed in the future, but I was thinking that's probably out-of-scope for this PR which is already getting large. So for now I just throw an exception.

**Comment by makaimann** [comment]
Here are the latest results. I think errors are mostly taken care of except for a few here and there for engines using interpolation. We're still not doing as well as `msat-ic3ia` with the native `ic3ia` implementation, although it gets better if we use `msat` or `boolector` with logging (shown in the `*-manual-logging-btor*` runs). I'm thinking we're about ready to merge and we can make incremental improvements (such as using actual sequence interpolants instead of an outer loop) in separate PRs. At this point we're comparable to master so performance can be a separate issue.

What do you think @ahmed-irfan ?

latest-param-ic3

**Comment by ahmed-irfan** [comment]
I think we can merge it and do the improvements in separate PRs.

I point to note is the number of memouts when we are using the logging solver from  smt-switch.

**Review by ahmed-irfan — APPROVED** [review]
Let's merge this and work of the improvements in separate PRs.

**Comment by makaimann** [comment]
Sounds good! I agree, this is getting large and it will be easier to make small improvements asynchronously without all the work being on one branch.

**Comment by makaimann** [comment]
For future reference, here were the `ic3ia` runs from the master branch before merging this PR. I.e. this refactor actually fixed some issues in `ic3ia` and improved performance, so we're headed in the right direction. 

master-ic3ia

**Pull request merged**
